06_gpu_and_ml/comfyui/comfyapp.py

```diff
@@ -72,126 +48,105 @@
+    .add_local_file(  # copy ComfyUI workflow JSON to the container
         Path(__file__).parent / "workflow_api.json",
         "/root/workflow_api.json",
         copy=True,
     )
 )
-# ### Downloading custom nodes
-# We'll use `comfy-cli` to download custom nodes, in this case the popular WAS Node Suite pack.
+# ## Downloading custom nodes
+# We'll also use `comfy-cli` to download custom nodes, in this case the popular [WAS Node Suite](https://github.com/WASasquatch/was-node-suite-comfyui).
+#
+# Use the [ComfyUI Registry](https://registry.comfy.org/) to find the specific custom node name to use with this command.
 image = (
     image.run_commands(  # download a custom node
         "comfy node install pr-was-node-suite-comfyui-47064894"
     )
     # Add .run_commands(...) calls for any other custom nodes you want to download
 )
 
-# See [this post](/blog/comfyui-custom-nodes) for more on how to install custom nodes on Modal.
-# ### Downloading models
-
-# You can also use comfy-cli to download models, but for this example we'll download the Flux models directly from Hugging Face into a Modal Volume.
-# Then on container start, we'll mount our models Volume into the ComfyUI models directory.
-# This allows us to avoid re-downloading the models every time you rebuild your image.
+# See [this post](/blog/comfyui-custom-nodes) for more examples on how to install popular custom nodes like [ComfyUI Impact Pack](/blog/comfyui-custom-nodes#2-comfyui-impact-pack) and [ComfyUI IPAdapter Plus](/blog/comfyui-custom-nodes#3-comfyui-ipadapater-plus).
+# ## Downloading models
 
-image = (
-    # install huggingface_hub with hf_transfer support to speed up downloads
-    image.pip_install("huggingface_hub[hf_transfer]==0.26.2")
-    .env({"HF_HUB_ENABLE_HF_TRANSFER": "1"})
-    .run_commands(  # needs to be empty for Volume mount to work
-        "rm -rf /root/comfy/ComfyUI/models"
-    )
-)
+# `comfy-cli` also supports downloading models, but we've found it's faster to use [hf_hub_download](https://huggingface.co/docs/huggingface_hub/en/guides/download#download-a-single-file) directly by:
+# 1. Enabling [faster downloads](https://huggingface.co/docs/huggingface_hub/en/guides/download#faster-downloads)
+# 2. Mounting the cache directory to a [Volume](/docs/guide/volumes)
+#
+# By persisting the cache to a Volume, we avoid re-downloading the models every time you rebuild your image.
 
-# We create the app and specify the image we built above.
 
-app = modal.App(name="example-comfyui", image=image)
+def hf_download():
+    from huggingface_hub import hf_hub_download
 
-#
-# First we need to run a function to download the Flux models to a Modal Volume.
+    flux_model = hf_hub_download(
+        repo_id="Comfy-Org/flux1-schnell",
+        filename="flux1-schnell-fp8.safetensors",
+        cache_dir="/cache",
+    )
 
-vol = modal.Volume.from_name("comfyui-models", create_if_missing=True)
+    # symlink the model to the right ComfyUI directory
```

Comment:
Hm, why are we symlinking in the `hf_download` function that is run as an image layer? Shouldn't we only need so symlink it in the `@enter` method of the function?

Reply:
i did it in the `hf_download` function because `hf_hub_download` returns the filepath where the model is saved so wanted to do the symlink within the same scope if that makes sense?